### ISSUE TITLE
Improve travis cfg

### ISSUE THREAD

**Issue by amerlyq** [albfan/ag.vim#24]
If I understand correctly, we can't make PR to add travis support into rking branch with no efforts?
Because it requires manual labour from rking itself.
However we could try to PR template configs and write issue with instructions?

Also, try to migrate your travis config to dockers (container) architecture, which runs much faster, but can't use sudo. Because `sudo` isn't actually necessary for this plugin testing.

**Comment by albfan** [comment]
I love docker so change it at your will.

For use travis rking just needs to login on travis with its github account and activate repo. But this is a custom testsuite so expect reluctant reactions proposing others to use it.

**Comment by amerlyq** [comment]
I think, it would be more fast for you to migrate `travis.yml` on your side.
Because while I will try it by myself, you will be spammed-out by broken builds:)

``` Yaml
sudo: false

addons:
  apt:
    packages:
    - vim
    - automake
    - pkg-config
    - libpcre3-dev
    - zlib1g-dev
    - liblzma-dev

before_install:
  - if [ ! -d "$HOME/ag" ]; then
    wget https://github.com/ggreer/the_silver_searcher/archive/master.zip;
    unzip master.zip && mv -vf the_silver_searcher-master ag;
    cd ag && ./build.sh && make install;
    fi

cache:
  directories:
  - $HOME/ag
```

Seems like it must be enough to run builds in dockers.
We build and cache ag, because it's not contained in approved sudo-less whitelist for travis/dockers.
However, if `make install` will work not as expected, you will need to add `export PATH="$HOME/ag:$PATH"`
Try it, please.

**Comment by albfan** [comment]
although sounds great, needing to compile ag is enough to refuse this change, meanwhile exist another solution which installs dependencies on OS side.

I would close this issue in the actual terms

**Comment by amerlyq** [comment]
My bad, I specified not full config, but only its part necessary to setup equivalent environment in dockers.
Therefore insert 'script:' section back into config, please.

**Comment by amerlyq** [comment]
About compiling `ag` -- I don't like it either. Currently building is necessary because no one added ag to white list of travis dockers packages. If someone will do it, we can simplify config much more.

Compiling thanks to `caching:` will be done only once in a while when dockers will be clearing cache. Therefore, as it will not be compiled on each commit, it's not that big deal -- especially as it takes dozen seconds at most. On other hands with dockers we have almost instant CI starting time and build status update, two times more system resources and newer underlying OS. Because current travis is not (and will not) upgrading sudo-required servers past Ubuntu 12.04. Only dockers will eventually do.

**Comment by amerlyq** [comment]
This reminds me about question I haven't asked yet.
For what purpose you use appveyor? Checking on windows?
I feel bad every time when look on all branches statuses and see only red crosses.
Is it possible to fix it or disable until fixed? Because currently overall status is irrelevant to actual code.
You may persist on lazy development and "don't fix until broken", but... it simply isn't nice, agree?

**Comment by amerlyq** [comment]
You erased my work by `push -f`. I'm too displeased, because I helped out to fix it.
Last state of travis.cfg:

``` Yaml
sudo: false

addons:
  apt:
    packages:
    - vim
    - automake
    - pkg-config
    - libpcre3-dev
    - zlib1g-dev
    - liblzma-dev

before_install:
  - if [ -d ./ag ]; then rm -rf ./ag; fi  # REM
  - if [ ! -d ./ag ]; then
    wget https://github.com/ggreer/the_silver_searcher/archive/master.zip;
    unzip master.zip && mv -vf the_silver_searcher-master ag; fi
  - if [ ! -f ./ag/ag ]; then
    (cd ./ag && ./build.sh)
    || (rm -rf ./ag; exit -1); fi
  - if [ ! -f ./ag/ag/ag- ]; then
    mv -v ./ag/ag{,-};
    printf '#!/bin/sh\nag- --noaffinity "$@"' > ./ag/ag;
    chmod +x ./ag/ag; fi
  - if ! hash ag >/dev/null 2>&1; then
    export PATH="$PWD/ag:$PATH"; fi

cache:
  directories:
  - $HOME/ag

script:
  - vim --version
  - ag --version
  - cd t && ./suite.sh --verbose --clean
```

**Comment by amerlyq** [comment]
Please, on future -- even if you don't like senseless commits which fix travis errors, commit them all, and only after all works, squash and rewrite history _once_. In all other cases you make unforgivable things.
And beside that, all your problems which you fight after push -f was already fixed. Want do it yourself?

**Comment by amerlyq** [comment]
And it seems, referring issues in commits which you constantly rewrite is bad idea, looking at this thread...

**Comment by albfan** [comment]
I did expect you to end this. As you ask me to complete, I didn't expect any work from you. Let's check your changes

**Comment by albfan** [comment]
I'm done. For me this change only adds complexity without noticiable benefit. My suggestion is to keep actual config for travis.

**Comment by amerlyq** [comment]
I know that I left task for you. I only simplified somehow your work, and it's all.
It not means I would completed all the work and closed issue.
But anyway good exuse for not checking repo state before push don't exist...

**Comment by amerlyq** [comment]
Maybe you are right, maybe not. 
This complexity is consequence from current state of travis infrastructure.
As I said before, it will be simplified as-as.
Besides, on travis -- dockers are recommended way. What you used is legacy.
Sudo-ers are kept only for projects not migrated yet or which can't migrate until sudo will be allowed in dockers on travis (planned).
I don't think you will look at configured travis in nearest future at all, until it works -- therefore its comlexity has not so much impact. 
In return we have latest upstream `ag` version to launch our tests, and it's good thing for those on bleeding-edge distros.

**Comment by albfan** [comment]
Ok to all, except one point: by now it's not working. Let's change this when we have it running.

About appveyor I feel to is annoying to have it just to say it fails. I will disable it by now

**Comment by amerlyq** [comment]
Don't be so impatient. I have pressed 'u' by accident in vim when selected snippet to copy in this window.
I was depressed at that moment after all.

All works after fixing those minor issues.

**Comment by amerlyq** [comment]
By the way -- this issue was opened 12 days ago.
It was right thing to ask you to complete this issue that time, when I had been accustoming to workflow and refactored main code having that passion. But now when I have for this issue some spare time and mood, ignited by your act of `docker` branch creation, I decided to make ease your life, because you was forced onto this issue back then. It was origin of all this contradiction in my words and actions.

**Comment by albfan** [comment]
I really appreciate simplicity, but if this is what travis offers to use it's latest interface, let's use it. One change to take in consideration is to add some env about ag version, to check plugin against important changes on ag, or to be sure stable versions of ag work with plugin (or the plugin tweaks depending on ag version).

it's not neccesary to implement right now, just remember it when compatibility problems arise. (i.e. you can merge or rebase (or rebase squashed) this, close issue and open a new issue about this enhancement)

**Comment by amerlyq** [comment]
Don't know what you mean by "env about ag version", because `ag --version` is already in travis.yml. I hope you will not `push -f` when I rebase-squash this;)

**Comment by albfan** [comment]
Hahaha, lesson learned. This is not anymore a onr man project, I will check.

Using 

```
env:
    - AG_VERSION=0.31
    - AG_VERSION=0.25
```

you get two different builds for travis. Just need to adapt wget to pick right ag source version.

```
wget http://blablabla${AG_VERSION}.zip
```

Other env variables can tweak plugin execution as needed

**Comment by amerlyq** [comment]
I liked your idea with build matrix. Good suggestion.
Therefore, as we build ag anyways, it was not that difficult to implement.
I squashed all and deleted branch. Enjoy:)
